[WFLY-12190] EJB: EJB over HTTP discovery configuration

tadamski opened the pull request:
https://issues.redhat.com/browse/EAP7-1236
https://issues.redhat.com/browse/WFLY-12190

Depends on:
https://github.com/wildfly/wildfly/pull/13305

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:
- [ ] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue(s)
- [ ] Pull Request contains description of the issue(s)
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [ ] Pull Request requires a change to the documentation
- [ ] Documentation have been updated accordingly
- [ ] Tests were added to cover changes

For new features ensure as well:
- [ ] Analysis was done
- [ ] Test Plan has been done
- [ ] Tests were verified in advance

If you are not an active contributor of the WildFly project you can request sponsorship by one of the members to help guide you through the process.

tadamski committed:
[WFLY-12190] EJB over HTTP discovery: configuration

darranl commented on ejb3/src/main/resources/org/jboss/as/ejb3/subsystem/LocalDescriptions.properties:
```diff
@@ -341,6 +341,10 @@ remoting-ejb-receiver.add=Adds a remoting ejb receiver reference to the profile
 remoting-ejb-receiver.remove=Removes a remoting ejb receiver reference from the profile
 remoting-ejb-receiver.outbound-connection-ref=Name of outbound connection that will be used by the ejb receiver
 remoting-ejb-receiver.connect-timeout=Remoting ejb receiver connect timeout
+remote-http-connection=TODO
```
TODO

darranl reviewed (changes requested):
The model description constants need defining.

However as on the analysis before we add a new server to server protocol we need to review the security implications.

tadamski commented:
@darranl I have updated the description. I have also added the possibility of HTTP discovery from standalone client (in attached PRs). Regarding security, EJB over HTTP protocol is integrated with Elytron and the invocations are secured. Specifically, I had to provide Elytron configuration for the tests to pass (discovery invocations to succeed). I agree that it has to be double-checked though.

darranl commented:
@tadamski I know the server side is already covered by Elytron security, the bit I am not sure about is the client side of these server to server connection.  In this case there are a few scenarios that may need some integration: -

- SSLContext
- Authentication Configuration 
- Identity Propagation using Delegated Credential

tadamski commented:
@darranl OK, I will try to investigate those but I would probably need some help. OTOH server->server invocations are already working in WildFly (with static discovery) so we may need to double-check that those are fine as well.

bstansberry commented:
@tadamski #12891 no longers includes the http-client part of this; it's just a bug fix update. The http-client part of this is now in a 1.1 release so this PR should include the upgrade to that.

pferraro reviewed (changes requested):
This change requires a model version bump, as well as transformers and transformation tests.

darranl commented:
@tadamski we haven't really followed up on this together, it looks like it has a merge conflict anyway but do we want to set something up so we can look at the security implications in more detail?

tadamski commented:
@darranl I have to update this PR but this is mostly config stuff. The key PR is https://github.com/wildfly/wildfly-http-client/pull/29 on which this PR relies. I would be happy to discuss it.

luck3y commented:
test, please ignore

tadamski committed:
[WFLY-12190] EJB over HTTP discovery: EjbOverHttpTestCase, EjbOverHttpDescriptorTestCase

tadamski committed:
[WFLY-12190] EJB over HTTP discovery: community docs

jamezp commented:
I've removed the missing-reqs label as there was an [agreement ](https://issues.redhat.com/browse/EAP7-1236?focusedCommentId=14105599&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14105599)the test development can be late on this feature.

fl4via commented:
This PR depends on https://github.com/wildfly/wildfly/pull/13305

Commit:
[WFLY-12190] EJB over HTTP discovery: EjbOverHttpWrongCredentialsTestCase

tadamski committed:
[WFLY-13492] Upgrade Wildfly HTTP Client to 1.1.2.Final

tadamski commented:
@jamezp merged; failing test seem unrelated; this should be ready to merge

zhantemirov commented:
I see all checks passing, do anything else prevent this PR from merging? Thanks

bstansberry commented on ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/jboss/as/ejb3/main/module.xml:
```diff
@@ -112,10 +112,13 @@
         <module name="org.hibernate.validator" optional="true"/>
         <module name="org.hibernate.validator.cdi" optional="true"/>
+
+        <module name="org.wildfly.http-client.ejb"/>
```
This is a duplicate.

tadamski replied:
fixed

bstansberry commented on ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java:
```diff
@@ -46,6 +48,7 @@
     private int defaultCompression;
 
     private Map<String, RemotingReceiverConfiguration> remotingReceiverConfigurations = new HashMap<String, RemotingReceiverConfiguration>();
+    private List<HttpConnectionConfiguration> httpConnectionConfigurations = new ArrayList<>();
```
This field should be final.  The ones above and below as well.

bstansberry replied:
Also, there is no code that modifies this list, so something is missing somewhere.

tadamski replied:
This was a bug. I have added the code that allows to configure http connections from jboss-ejb-client.xml file and added a test for that EjbOverHttpDescriptorTestCase

bstansberry commented on ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java:
```diff
@@ -346,4 +353,16 @@ public int hashCode() {
         }
     }
+
+    public class HttpConnectionConfiguration {
```
This should be final and presumably static.

tadamski replied:
fixed

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+[[profile]]
+=== <profile>
+
+Remote profile specificies a configuration of remote invocations that can
```
s/Remote profile/A remote profile/g
s/specificies/specifies

tadamski replied:
fixed

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+
+Remote profile specificies a configuration of remote invocations that can
+be referenced by many deployments. EJBs that are meant to be invoked can
+be discovered in either static or dynamic way.
```
s/either static or dynamic/either a static or a dynamic/g

tadamski replied:
fixed

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+be referenced by many deployments. EJBs that are meant to be invoked can
+be discovered in either static or dynamic way.
+
+Static discovery decides which remote node to connect based on the information
```
s/to connect/to connect to/g

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+Static discovery decides which remote node to connect based on the information
+provided by the user.
+
+Dynamic discovery is responsible for monitoring available EJBs on all the
```
s/monitoring avaiable/monitoring the available/g

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+
+Dynamic discovery is responsible for monitoring available EJBs on all the
+nodes to which connections are configured and decides which remote node to
+connect based on gathered data.
```
"connect to based on the gathered data."

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+be discovered in either static or dynamic way.
+
+Static discovery decides which remote node to connect based on the information
+provided by the user.
```
Perhaps s/user/administrator  ?

AIUI the end user making the request isn't involved, so it's a user who administers the server config.

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+[[static-ejb-discovery]]
+==== <static-ejb-discovery>
+
+Static ejb discovery allows to explicitly specify on which remote nodes
```
s/allows/allows the administrator/g

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+==== <static-ejb-discovery>
+
+Static ejb discovery allows to explicitly specify on which remote nodes
+given EJBs are located. `module` tag is used to define it:
```
s/`module`/The `module`/g

tadamski replied:
fixed

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+[[remoting-ejb-receiver]]
+==== <remoting-ejb-receiver>
+
+`remoting-ejb-receiver` tag is used to define dynamic discovery based on
```
s/`remoting-ejb-receiver`/The `remoting-ejb-receiver`/g

tadamski replied:
fixed

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+==== <remoting-ejb-receiver>
+
+`remoting-ejb-receiver` tag is used to define dynamic discovery based on
+the remoting procol:
```
s/procol/protocol

tadamski replied:
fixed

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+[[remote-http-connection]]
+==== <remote-http-connection>
+
+`remote-http-connection` tag is used to define dynamic discovery based on
```
s/`remote-http-connection`/The `remote-http-connection`/g

tadamski replied:
fixed

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:
```diff
@@ -174,6 +184,51 @@ This is used to enable remote EJB invocations. It specifies a space-separated li
+==== <remote-http-connection>
+
+`remote-http-connection` tag is used to define dynamic discovery based on
+HTTP procol:
```
s/procol/protocol

tadamski replied:
fixed

darranl replied:
I think the main question I would like to ask is - is there sufficient test coverage of clients using authentication client configuration for a dynamically discovered destination?

As mentioned earlier we have aspects such as SSL and the credentials for authentication which should already be covered by the EJB client to verification that the behaviour is correct.  In this case I believe it will be preferable for the resolved configuration to be based on the discovered destination.

tadamski replied:
The test coverage for the dynamically discovered destinations would require additional analysis. OTOH, dynamic discovery is main route of invocation also for remoting and this commit just adds this feature for HTTP. Currently EAP/WildFly already relies on dynamic invocation and if we decide that the testsuite have to be extended it should be for done a number of scenarios, possibly independently of this PR.

darranl replied:
Ok, lets go with what we have here - if you are happy HTTP is just using the same pattern as the existing approaches we should be good.

tadamski replied:
Hmm, but just to clarify cause I'm not sure if I understand correctly. Discovery provider and HttpEjbReceiver (which is used to perform invocation to discovered destination) (also other receivers) resolve their AuthenticationContexts independently:
https://github.com/wildfly/wildfly-http-client/blob/master/ejb/src/main/java/org/wildfly/httpclient/ejb/HttpEJBReceiver.java#L132

Should this be modified?

bstansberry reviewed (changes requested):
I've made some comments. 

For this one to move forward we should get an approval from @darranl.

bstansberry commented:
/retest

bstansberry commented:
The issues I've raised have all been addressed, but the discussion with Darran remains open.

darranl commented on testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/ejb/http/EjbOverHttpWrongCredentialsTestCase.java:
```diff
@@ -0,0 +1,85 @@
+package org.jboss.as.test.multinode.ejb.http;
```
Missing header

tadamski replied:
fixed

darranl commented:
I have one comment for testing that I don't think should hold up this PR so is something more general.  At the moment our testsuite is becoming very slow to execute so the pattern of Configure -> Deploy -> Single Test -> Undeploy -> Revert Configuration adds a lot of steps for one test.  Ideally if we can find ways a common deployment can be used for multiple scenarios we can execute multiple tests for a single pair or configure and pair of deploy tasks.  Ideally finding ways to test the success and failure scenarios under the same test case.

But in the case of this specific change I see we have some simple authentication client configuration where a single configuration is available but I would still like to see something where we test a configuration where the correct configuration is only resolved / resolvable after discovery.  I don't necessarily have an issue with that being in a follow up PR if we can make sure it happens.

The reason I am seeing this piece as important is because it will be an area of code we must maintain for backwards compatibility so once we release a Final version we will not be able to change the resolution logic without the risk of breaking something developed against the original behaviour.

bstansberry reviewed (changes requested):
@tadamski This needs an update to the transformers to the previous management API version:

https://github.com/wildfly/wildfly/blob/master/ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java#L289

Ejb3SubsystemUnitTestCase and Ejb3TransformersTestCase should cover it as well.

bstansberry commented:
@tadamski https://issues.redhat.com/browse/WFLY-13878 is to track my last request.

chengfang commented:
@bstansberry I'll update the transformer to handle the new element `remote-http-connection`
